cpp/src/join/quadtree_poly_filtering.cu

```diff
@@ -0,0 +1,292 @@
+                                                               num_pairs,
+                                                               x_min,
+                                                               y_min,
+                                                               scale,
+                                                               max_depth,
+                                                               stream);
+    } else {
+      // Otherwise, test the current level for intersections
+      std::tie(num_parents, num_children) = find_intersections(quadtree,
+                                                               poly_bbox,
+                                                               tmp_node_idxs.begin(),
+                                                               tmp_poly_idxs.begin(),
+                                                               tmp_pairs,
+                                                               out_pairs + num_results,
+                                                               num_pairs,
+                                                               x_min,
+                                                               y_min,
+                                                               scale,
+                                                               max_depth,
+                                                               stream);
+    }
+
+    num_results += num_children;
+
+    // stop descending if no parent quadrants left to expand
+    if (num_parents == 0) break;
+
+    // Shrink the intermediate level storage buffers to overwrite removed elements
+    tmp_types.shrink_to_fit(stream);
+    tmp_levels.shrink_to_fit(stream);
+    tmp_node_idxs.shrink_to_fit(stream);
+    tmp_poly_idxs.shrink_to_fit(stream);
+
+    // Use the current parent quad indices as the element indices lookup for the global child counts
+    auto child_counts =
+      thrust::make_permutation_iterator(node_counts.begin<uint32_t>(), tmp_node_idxs.begin());
+
+    // Walk one level down and fill the intermediate storage buffers with the next level's poly and
+    // quad indices
+    auto next_level = descend_quadtree(child_counts,
+                                       node_offsets.begin<uint32_t>(),
+                                       num_parents,
+                                       tmp_types,
+                                       tmp_levels,
+                                       tmp_node_idxs,
+                                       tmp_poly_idxs,
+                                       stream);
+
+    num_pairs     = std::get<0>(next_level);
+    tmp_types     = std::move(std::get<1>(next_level));
+    tmp_levels    = std::move(std::get<2>(next_level));
+    tmp_node_idxs = std::move(std::get<3>(next_level));
+    tmp_poly_idxs = std::move(std::get<4>(next_level));
+    // update tmp_pairs iterator to get ready for next level iteration
+    tmp_pairs = make_zip_iterator(
```

Comment:
Alternatively, just pass `next_level` directly to `make_zip_iterator`, since it takes a tuple, either way eliminates all the `move` calls.